### PR TITLE
Add cluster configuration to allow limiting the number of multi-stage queries running concurrently

### PR DESCRIPTION
**Pull request by yashmayya**
- Currently, there is no limit on the number of multi-stage queries that are executed concurrently. Every request that comes in to a broker is immediately compiled into a query plan and dispatched to the servers for query execution.
- On the servers, each stage in the query plan is submitted to a cached thread pool executor. This means that we don't have an effective way to limit the number of threads that are being utilized to execute multi-stage queries on servers. This can result in a very large number of multi-stage query executor threads in high QPS / complex query workloads leading to resource contention and performance degradation.
- Simply using a fixed thread pool executor instead is not a good solution because this could lead to distributed deadlocks and query timeouts / failures. Similarly, it isn't trivial to implement the max concurrent query execution logic in the servers without considering global ordering.
- Consider this example from @gortiz - 
  - Given a query `Q1` is formed by stages `S11` and `S12`, where `S11` depends on data from `S12`.
  - And a query `Q2` is formed by stages `S21` and `S22`, where `S21` depends on data from `S22`
  - If `Q1` and `Q2` arrive at the same time and:
    - Server `Srv1` receives `Q1` first, starts executing `S11` and ends up waiting for server `Srv2` to execute `S12`
    - Server `Srv2` receives `Q2` first, starts executing `S21` and ends up waiting for server `Srv1` to execute `S22`
    - Server `Srv1` receives `Q2` and would execute `S22`, but sees that it has reached max number of concurrent queries, so it waits.
    - Server `Srv2` receives `Q1` and would execute `S12`, but sees that it has reached max number of concurrent queries, so it waits.
    - This leads to a situation where both the queries are blocked and will eventually timeout and fail.
- This patch takes an alternative approach of limiting the number of concurrently executing multi-stage queries in the broker itself.
- A new cluster level config is introduced to limit the number of multi-stage queries running concurrently. The config is a "per server" value and can be potentially re-used in the future if / when we move to server side throttling. This is disabled by default and is currently an opt-in throttling mechanism.
- If set to a value > 0, the value will be divided among all the brokers - i.e., if the config value is set to `6` and there are 2 brokers and 3 servers in the cluster, each broker will be able to execute 9 multi-stage queries concurrently. If there are already 9 queries currently executing on a broker, any subsequent incoming queries will wait for a slot to open up.
- This is implemented using a custom semaphore which reacts to changes in the cluster config value or the number of brokers / servers in the cluster in order to recalculate the total available permits. A custom semaphore implementation is needed here because we want to use the [Semaphore::reducePermits](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Semaphore.html#reducePermits-int-) method (when # of brokers increases, # of servers decreases, or the cluster config value for max concurrent queries decreases) which has `protected` access. We need to use this method in order to be able to reduce the total number of permits for the semaphore without blocking.
- Note that this first cut implementation is fairly primitive and assumes that queries are distributed evenly across the brokers and that each query will be executed on all the servers. Alternatively, we could check the actual servers that a query touches or even the number of workers to estimate the number of threads a query uses per server, but this would introduce additional complexity from both the user and developer POV.

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    _maxConcurrentQueries = maxConcurrentQueries;
+    _numBrokers = Math.max(1, numBrokers);
+    _totalPermits =
+        maxConcurrentQueries > 0 ? Math.max(1, maxConcurrentQueries / Math.max(numBrokers, 1)) : maxConcurrentQueries;
+  }
+
+  @Override
+  public void acquire()
+      throws InterruptedException {
+    // If _totalPermits is <= 0, it means that the cluster is not configured to limit the number of multi-stage queries
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java:62]: It is fine to keep it here, but given this is an important attribute invariant, I think it is worth to add it as javadoc in the attribute itself.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  private int _maxConcurrentQueries;
+  private int _totalPermits;
+
+  public MultiStageQuerySemaphore(int numBrokers, int maxConcurrentQueries) {
+    super(Math.max(1, maxConcurrentQueries / Math.max(numBrokers, 1)), true);
+    _maxConcurrentQueries = maxConcurrentQueries;
+    _numBrokers = Math.max(1, numBrokers);
+    _totalPermits =
+        maxConcurrentQueries > 0 ? Math.max(1, maxConcurrentQueries / Math.max(numBrokers, 1)) : maxConcurrentQueries;
+  }
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java:57]: I think we are using too many _max_es here. IMO it would be better to create a static factory method that does all the max logic and then call a simple constructor.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class helps limit the number of multi-stage queries being executed concurrently. Currently, this limit is
+ * applied at the broker level, but could be moved to the server level in the future.
+ */
+public class MultiStageQuerySemaphore extends Semaphore implements ClusterChangeHandler {
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java:40]: IMHO this class does too many things. I would suggest to split it in two different classes:

1. The semaphore itself. It doesn't need to know the number of brokers, max concurrent queries, etc. It has a `setTotalPermits` for a new number of permits, which calls `super.reducePermits` and `super.release` if needed.
2. The ClusterChangeHandler, which has a reference to the semaphore. It knows the number of brokers and max concurrent queries and calls `setTotalPermits` each time there are cluster changes.

By doing so the semaphore has no knowledge related to this feature and it is just a semaphore with the ability to change permits. It is easier to understand and we can use it in other developments. And different handlers can be used as needed.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/BaseBrokerStarter.java [hunk 4]
```diff
@@ -335,13 +339,26 @@ public void start()
...
     QueryDispatcher queryDispatcher = null;
     if (_brokerConf.getProperty(Helix.CONFIG_OF_MULTI_STAGE_ENGINE_ENABLED, Helix.DEFAULT_MULTI_STAGE_ENGINE_ENABLED)) {
+      int numBrokers = Math.max(1, (int) _helixAdmin
+          .getInstancesInCluster(_spectatorHelixManager.getClusterName())
+          .stream()
+          .filter(instance -> instance.startsWith(CommonConstants.Helix.PREFIX_OF_BROKER_INSTANCE))
+          .count());
+      int maxConcurrentQueries = Integer.parseInt(
+          _helixAdmin.getConfig(new HelixConfigScopeBuilder(HelixConfigScope.ConfigScopeProperty.CLUSTER).forCluster(
+                      _spectatorHelixManager.getClusterName()).build(),
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/BaseBrokerStarter.java:349]: I don't think this code belongs here. It should be done automatically by `MultiStageQuerySemaphore`

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java (hunk 3 above)
Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java:40]: Probably you can even generalize the handler to depend on some function callbacks or abstract methods so you don't need to mock all the helix attributes. I would prefer that (the less mocking the better) but I understand it can also be seen as overengineering.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java [hunk 5]
```diff
@@ -224,67 +227,89 @@ protected BrokerResponse handleRequest(long requestId, String query, SqlNodeAndO
...
+      if (!_querySemaphore.tryAcquire(queryTimeoutMs, TimeUnit.MILLISECONDS)) {
+        LOGGER.warn("Timed out waiting to execute request {}: {}", requestId, query);
+        requestContext.setErrorCode(QueryException.EXECUTION_TIMEOUT_ERROR_CODE);
+        return new BrokerResponseNative(QueryException.EXECUTION_TIMEOUT_ERROR);
       }
-      LOGGER.warn("Timed out executing request {}: {}", requestId, query);
+    } catch (InterruptedException e) {
+      LOGGER.warn("Interrupt received while waiting to execute request {}: {}", requestId, query);
       requestContext.setErrorCode(QueryException.EXECUTION_TIMEOUT_ERROR_CODE);
       return new BrokerResponseNative(QueryException.EXECUTION_TIMEOUT_ERROR);
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java:242]: Maybe not in this PR, but we will probably need to have a way to tell the user where we were waiting most of the time. It could be useful for example to tell users that their queries are timing out because other queries are consuming all threads.

**Review by gortiz — COMMENTED** [review]
Added some comments, mainly on code style. Anyway we need to discuss whether we want to use these config options or some other

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+import org.apache.pinot.broker.broker.helix.ClusterChangeHandler;
+import org.apache.pinot.common.concurrency.AdjustableSemaphore;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class helps limit the number of multi-stage queries being executed concurrently. Currently, this limit is
+ * applied at the broker level, but could be moved to the server level in the future. Note that the cluster
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:40]: nit: This javadoc says twice that we could move the check to server level.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+ * configuration is a "per server" value and the broker currently simply assumes that a query will be across all
+ * servers. Another assumption here is that queries are evenly distributed across brokers. Ideally, we want to move to a
+ * model where the broker asks each server whether it can execute a query stage before dispatching the query stage to
+ * the server. This would allow for more fine-grained control over the number of queries being executed concurrently
+ * (but there are some limitations around ordering and blocking that need to be solved first).
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:45]: nit: I think we are merging two different things here. In my mind there are 3 options:

1. Broader broker check, the one we are implementing, which simplifies the check by assuming evenly distributed queries across brokers and servers.
2. Ask servers if they can execute queries, where the broker ask the servers whether they can execute more queries or not (probably this should be done while workers/plan fragments are assigned). A server needs to be able to execute all the associated workers of the query atomically. This means that a query that spawns a lot of workers on a single server may consume a lot of threads in the server.
3. Ask servers if they can execute workers. This is what implies there could be blocks if we don't care about ordering.

The difference between 2 and 3 is that queries that require tons of workers may end up consuming a lot of threads in 2, while in 3 we would execute only the some workers and once they finish the other will start to run. In order to do so we need to execute the dag in a consistent manner (and probably buffer some data).

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(MultiStageQueryThrottler.class);
+
+  private HelixManager _helixManager;
+  private HelixAdmin _helixAdmin;
+  private HelixConfigScope _helixConfigScope;
+  private int _numBrokers;
+  private int _numServers;
+  // If _maxConcurrentQueries is <= 0, it means that the cluster is not configured to limit the number of multi-stage
+  // queries that can be executed concurrently. In this case, we should not block the query.
+  private int _maxConcurrentQueries;
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:58]: nit: It would be better to set this as javadoc `/**` to be able to use IDE intregrations

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        // This operation isn't safe to do while queries are running so we require a restart of the broker for this
+        // change to take effect.
+        LOGGER.warn("Enabling or disabling limitation of the maximum number of multi-stage queries running "
+            + "concurrently requires a restart of the broker to take effect");
+        return;
+      }
+
+      if (maxConcurrentQueries > 0) {
+        _semaphore.setPermits(Math.max(1, maxConcurrentQueries * _numServers / _numBrokers));
+      }
```
Comment (gortiz) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:159]: I think there is a possible NPE here. Imagine that the broker is initialized with -1 concurrent queries, `_semaphore` is kept as null. Then the config is changed to set an actual max concurrent queries, then we try to call `_semaphore.setPermits` here without actually creating the semaphore.

For simplicity, I would suggest to always allocate the semaphore (with whatever permits). Given don't delegate on the semaphore when `_maxConcurrentQueries <= 0`, that should be a clean and correct solution

Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:159]: Maybe we can add a test to verify my scenario.

File: pinot-common/src/main/java/org/apache/pinot/common/concurrency/AdjustableSemaphore.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    super(permits);
+    _totalPermits = permits;
+  }
+
+  public AdjustableSemaphore(int permits, boolean fair) {
+    super(permits, fair);
+    _totalPermits = permits;
+  }
+
+  public void setPermits(int permits) {
```
Comment (gortiz) [inline pinot-common/src/main/java/org/apache/pinot/common/concurrency/AdjustableSemaphore.java:41]: Just to fail fast (and in controlled way) I would check that permits > 0

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java (hunk 5 above)
Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java:242]: a metric would be great, a log could be useful. But I was more worried about immediate the feedback to the user. I mean, the error information we return. But it was a wishful thinking more than an actual requirement.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java (hunk 9 above)
Comment (yashmayya) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:159]: > I think there is a possible NPE here. Imagine that the broker is initialized with -1 concurrent queries, _semaphore is kept as null. Then the config is changed to set an actual max concurrent queries, then we try to call _semaphore.setPermits here without actually creating the semaphore.

Unless I'm misunderstanding you, that condition is already being checked just before this line and we're disallowing such a change right? This makes things a lot more simple to reason about - transitioning from disabled to enabled or vice versa while queries are running will create lots of edge cases to handle with the acquire / release mechanism.

> Maybe we can add a test to verify my scenario.

There's already tests verifying disabled -> enabled and enabled -> disabled transitions.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java (hunk 7 above)
Comment (yashmayya) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:45]: Makes sense, I was assuming that for the 2nd option the check would happen before dispatch instead of during planning. Either way, this is just a rough sketch for a future idea so I've just removed this part of the Javadoc to avoid confusion. This PR can always be re-visited for more context in the future.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java (hunk 9 above)
Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:157]: Oh! I didn't _parse_ the previous check. It makes sense. In that case queries that were running when the feature was disabled will suddenly try to release the permit once they finish.

I guess we can resolve that by keeping a reference to a virtual _semaphore_ (not an actual Semaphore, but something that implements the API) and then call release on that object, which will be a noop on queries started when the feature was disabled.

But it is true that it is not a blocking issue and we can assume it requires a restart in this case.

But... isn't that actually going to be problematic when moving from N to M permits? I mean:

1. The semaphore is configured with N (where N > 0) permits.
1. Query starts, semaphore has N (where N > 0) permits, so it consumes one and continues. The semaphore has now N-1 permits.
2. The config changes and now we need to use use M permits. We restart the semaphore to use that number.
3. The query ends and release its permit. Now the semaphore has M+1 permits.

In the worst case scenario we could end up having N+M permits. Luckily, this is not accumulative. But unless I'm wrong I think we need to fix the issue.

Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:157]: I have two proposals to fix it:
1. Modify `tryAcquire` to return an object that:
  * has a method to check whether we are allowed to go
  * is AutoCloseable. Its close method release the lock if and only if the semaphore didn't change (we can use a version stored in a AtomicInt to verify that without locking).
2. Keep in this object a set of threads and modify the release method to only release the permit if the caller thread is listed in the set. We can clean the set when config changes. This requires less changes in the code, but the synchronization may be expensive.

In general I think we need to synchronize in this class anyway

Comment (yashmayya) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:157]: > But... isn't that actually going to be problematic when moving from N to M permits

Hmm, I don't think so - we're increasing / decreasing the number of permits based on the previous total number of permits, not the current available number of permits. You can see some of the unit tests like `testIncreaseNumBrokers` to see how this would look like in action.

> In general I think we need to synchronize in this class anyway

The cluster change handling should be done in a single threaded fashion and the semaphore should support concurrent modification of the number of permits while other threads are trying to acquire / release so I don't think additional external synchronization is necessary here?

Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQueryThrottler.java:157]: You are right. I though we were calling a _set_ method to change the number of permits instead of sending the increase/decrease.

**Review by gortiz — APPROVED** [review]

**Comment by gortiz** [comment]
+1
